nav2_util/include/nav2_util/service_client.hpp

```diff
@@ -77,13 +99,22 @@ class ServiceClient
       node_->get_logger(), "%s service client: send async request",
       service_name_.c_str());
     auto future_result = client_->async_send_request(request);
-
-    if (callback_group_executor_.spin_until_future_complete(future_result, timeout) !=
-      rclcpp::FutureReturnCode::SUCCESS)
-    {
-      // Pending request must be manually cleaned up if execution is interrupted or timed out
-      client_->remove_pending_request(future_result);
-      throw std::runtime_error(service_name_ + " service client: async_send_request failed");
+    if(use_internal_executor_) {
+      if (callback_group_executor_->spin_until_future_complete(future_result, timeout) !=
```

Comment:
Can the new `spin_until_complete` API added here be used instead? That handles the `if/else` on `use_internal_executor_`. This (and below) couldbe reduced to:

```
if (spin_until_complete(future_result, timeout) != rclcpp::FutureReturnCode::SUCCESS)
{
        // Pending request must be manually cleaned up if execution is interrupted or timed out
        client_->remove_pending_request(future_result);
        throw std::runtime_error(service_name_ + " service client: async_send_request failed");
}
```

This would also nicely remove the redundancy in the removing pending request and throwing code :smile: